File: back-end/src/main/scala/com/idak/tuto/api/service/UserService.scala
package com.idak.tuto.api.service

import com.idak.tuto.api.model.User
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

import scala.collection.mutable.ListBuffer

@Service
class UserService(passwordEncoder: PasswordEncoder) {
  var users = new ListBuffer[User]
  users += new User(generateId,"admin",passwordEncoder.encode("admin"),"admin@gmail.com")
  users += new User(generateId,"user",passwordEncoder.encode("user"),"user@gmail.com")



  def create(user: User): User = {
    user.setId(generateId)
    user.setPassword(passwordEncoder.encode(user.getPassword))
    users += user
    user
  }

  def delete(id: Int) = {
    users --= users.filter(_.getId == id)
  }

  def get = users.toArray

  def getOne(id: Int) : Option[User] = users.find(_.getId == id)

  def authenticate(username: String, password: String) : Option[User] = {
    users.find(u => u.getUsername.equals(username) && passwordEncoder.matches(password, u.getPassword))
  }

  private def generateId : Int = {
    if (users.isEmpty) 1 else users.map(_.getId).max + 1
  }


}

File: back-end-java/src/main/java/com/idak/tuto/api/service/UserService.java
package com.idak.tuto.api.service;

import com.idak.tuto.api.repository.UserRepository;
import com.idak.tuto.api.model.User;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * User Service implementation
 */
@Service
public class UserService {

    private UserRepository userRepository;
    private PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
        passwordEncoder = new BCryptPasswordEncoder();
    }

    /**
     * Create new User
     * @param user : user to create
     * @return User : saved user
     */
    public User create(User user) {
        user.setPassword(passwordEncoder.encode(user.getPassword()));
        return userRepository.save(user);
    }

    /**
     * Delete User
     * @param user
     */
    public void delete(User user) {
        userRepository.delete(user);
    }

    /**
     * Delete User by id
     * @param id
     */
    public void delete(int id) {
        userRepository.delete(id);
    }

    public List<User> get() {
        return (List<User>) userRepository.findAll();
    }

    /**
     * Find User by id
     * @param id
     * @return User
     */
    public User get(int id) {
        return userRepository.findOne(id);
    }

    /**
     * Update user
     * @param user
     * @return User
     */
    public User update(User user) {
        return userRepository.save(user);
    }

    /**
     *  Check username and password
     * @param username
     * @param password
     * @return Optional
     */
    public Optional<User> authenticate(String username, String password){
        if(Objects.isNull(username) || Objects.isNull(password)) {
            return Optional.empty();
        }

        Optional<User> authenticatedUse = userRepository.findByUsername(username);
        if(authenticatedUse.isPresent()
                && passwordEncoder.matches(password, authenticatedUse.get().getPassword())){
            return authenticatedUse;
        }

        return Optional.empty();

    }
}
